src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -140,38 +140,28 @@ internal void Freeze()
             _orderedSynthesizedMethods = _synthesizedMethods.OrderBy(kvp => kvp.Key).Select(kvp => kvp.Value).AsImmutable();
 
             // Sort proxy types.
-            _orderedProxyTypes = _proxyTypes.OrderBy(kvp => kvp.Key).Select(kvp => kvp.Value).AsImmutable();
+            _orderedProxyTypes = _proxyTypes.OrderBy(kvp => kvp.Key.Size).Select(kvp => kvp.Value).AsImmutable();
         }
 
         private bool IsFrozen => _frozen != 0;
 
-        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data)
+        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort alignment)
         {
             Debug.Assert(!IsFrozen);
-            Cci.ITypeReference type = _proxyTypes.GetOrAdd((uint)data.Length, GetStorageStruct);
-            return _mappedFields.GetOrAdd(data, data0 =>
-            {
-                var name = GenerateDataFieldName(data0);
-                var newField = new MappedField(name, this, type, data0);
-                return newField;
-            });
-        }
 
-        private Cci.ITypeReference GetStorageStruct(uint size)
-        {
-            switch (size)
-            {
-                case 1:
-                    return _systemInt8Type ?? new ExplicitSizeStruct(1, this, _systemValueType);
-                case 2:
-                    return _systemInt16Type ?? new ExplicitSizeStruct(2, this, _systemValueType);
-                case 4:
-                    return _systemInt32Type ?? new ExplicitSizeStruct(4, this, _systemValueType);
-                case 8:
-                    return _systemInt64Type ?? new ExplicitSizeStruct(8, this, _systemValueType);
-            }
+            Cci.ITypeReference type = _proxyTypes.GetOrAdd(
+                ((uint)data.Length, alignment), key =>
+                key.Size switch
+                {
+                    1 when _systemInt8Type is not null => _systemInt8Type,
+                    2 when _systemInt16Type is not null => _systemInt16Type,
+                    4 when _systemInt32Type is not null => _systemInt32Type,
+                    8 when _systemInt64Type is not null => _systemInt64Type,
+                    _ => new ExplicitSizeStruct(key.Size, key.Alignment, this, _systemValueType)
+                });
 
-            return new ExplicitSizeStruct(size, this, _systemValueType);
+            return _mappedFields.GetOrAdd((data, alignment), key =>
+                new MappedField(GenerateDataFieldName(key.Data, key.Alignment), this, type, key.Data));
```

Comment:
>new MappedField(GenerateDataFieldName(key.Data, key.Alignment), this, type, key.Data)) [](http://example.com/codeflow?start=16&length=86)

Does offset of this field in private implementation detail type make any difference?

Reply:
It looks like this field is static, so it is not a subject to layout within a type instance.

Reply:
I'll let @davidwrighton confirm, but my understanding is that the offset of the field doesn't matter. What matters is the alignment and .pack flag on the blob that the field points to.

Reply:
> It looks like this field is static, so it is not a subject to layout within a type instance.

The .pack is for the type of the field (not the type containing the field), and according to the docs "The alignment of the type is the size of its largest element (1, 2, 4, 8, etc., bytes) or the specified packing size, whichever is smaller."